rfcs/0000-nixpkgs-ifd.md

```diff
@@ -0,0 +1,145 @@
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+TODO, demonstrate changes to Nixpkgs, e.g. using the Haskell infrastructure, in a fork, and link?
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+1. > We've *doubled* the amount of evaluation we do, oh no!
+
+   Sounds scary, I know.
+   But I don't think that's bad, actually.
+   What's so bad today is the time and memory usage of *each* evaluation.
+
+   You can think of that as a bunch of ungainly massive rectangular tiles we are trying to fit on a floor, the floor being our machine resources to schedule.
+   What would be really bad is *increasing the tile size*.
+   This means we need a bigger floor or else laying tile is harder.
+   What this is doing is *increasing the number of times*.
+   We can simple add a second floor to solve any problems that arise from that.
+
+   This is simplification, sure, but I think the parable is correct to the "real" situation, too.
+
+2. > Isn't IFD really slow?
+
+   What is slow is that evaluator has no parallelism.
+   That means is that every time we hit an *unbuilt* derivation, we block until it's finished building.
+   Worse, even if it is easy to run, we're probably going to check some substituters, etc., so there are all sorts of slow IO round trips making the critical path worse.
+   We could fix this, but there is no energy to do so right now.
+   Making the evaluator parallel without making our memory issues worse is hard work.
+
+   But, none of that matters for this proposal.
+   `hydra.nixos.org` will only need to read built paths, and that shouldn't be meaningfully slower than regular `import`-ing.
+
+3. > IFD, is too controversial, don't do it!
+
+   I think this is a classic example of don't let the perfect be the enemy of the good.
+   The problems with IFD and the problems IFD is trying trying to address both don't let a lot of attention.
+   The fact of the matter is Nixpkgs is how this community coordinates with itself, and agrees on priorities.
+   If it isn't being used in Nixpkgs, there is hard ceiling of how much attention it will get.
+
+   The benefits of IFD don't get enough attention.
+   A package in Nixpkgs is more than a derivation: there's the `meta` as I mentioned above.
+   There's also being able to read the code and (somewhat) understand what's going in reference to other derivations.
+   Finally, there's being able to `override`, `overrideAttrs`, etc. the derivation downstream.
+   IFD alone allows computed packages that follow all these norms.
+
+4. Say I want IFD that depends on other IFD?
+
+   In other words, can one import a derivation that is itself evaluated from and import from a derivation?
+   No, not without introducing another round of building and evaluating for Hydra.
+   But I don't think we need arbitrarily-deep dynamism anyways: it is a tool that should be used with care anyways, because stasis \[staticism?\] is the goodly disciplinarian that makes Nixpkgs so great.
```

Comment:
Let's say we generate the Python packages set using IFD. There are packages there that are used a lot by other applications/tools, an example is Sphinx. Haskell's GHC depends on Sphinx. Will generating the Haskell packages set depend on the IFD used for Python? I suppose if the Haskell set generator is written in Haskell.

Reply:
Added alternative and example covering my point above in greater detail. Thanks.